src/EditorFeatures/Text/ITextImageHelpers.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Text;
+
+internal static class ITextImageHelpers
+{
+    private static readonly Func<int, int, string> s_textLog = (v1, v2) => string.Format("FullRange : from {0} to {1}", v1, v2);
+
+    private static readonly Func<ITextChange, TextChangeRange> s_forwardTextChangeRange = c => CreateTextChangeRange(c, forward: true);
+    private static readonly Func<ITextChange, TextChangeRange> s_backwardTextChangeRange = c => CreateTextChangeRange(c, forward: false);
+
+    public static IReadOnlyList<TextChangeRange> GetChangeRanges(ITextImage oldImage, ITextImage newImage)
+        => GetChangeRanges(oldImage.Version, newImage.Version);
+
+    public static IReadOnlyList<TextChangeRange> GetChangeRanges(ITextImageVersion oldImageVersion, ITextImageVersion newImageVersion)
+    {
+        var forward = oldImageVersion.VersionNumber <= newImageVersion.VersionNumber;
+
+        var oldSnapshotVersion = forward ? oldImageVersion : newImageVersion;
+        var newSnapshotVersion = forward ? newImageVersion : oldImageVersion;
+
+        INormalizedTextChangeCollection? changes = null;
+        for (var oldVersion = oldSnapshotVersion;
+            oldVersion != newSnapshotVersion;
+            oldVersion = oldVersion.Next)
+        {
+            if (oldVersion.Changes.Count != 0)
+            {
+                if (changes != null)
+                {
+                    // Oops - more than one "textual" change between these snapshots, bail and try to find smallest changes span
```

Comment:
>// Oops - more than one "textual" change between these snapshots, bail and try to find smallest changes span [](http://example.com/codeflow?start=20&length=108)

I know this comment isn't new in this PR, but I was trying to understand this code, and the "oops" part of this is confusing. I would think this would commonly happen if the versions sent in had some intermediary versions between.

Is it intended to instead call the GetChangeRanges method that accepts the forward parameter if you might have versions that aren't adjacent?

Reply:
i think we call the topmost helper.  and it then calls the lower level helper one a time through each version change.

Reply:
@jasonmalinowski  may know moer her.e